micronota/db/uniref.py

```diff
@@ -314,52 +220,55 @@ def prepare_metadata(in_fp, out_fp, append=True):
     The schema of the database file contains one table named `tigrfam` that
     has following columns:
 
-    1. ``ac``. TEXT. UniProtKB accession.
-
-    2. ``dataset``. TEXT.
+    1. ``ac``. TEXT. UniProtKB primary accession.
 
-    3. ``taxon``. TEXT.
+    2. ``status``. INT. The index in ``_status``
 
-    4. ``transfer``. INTEGER. Used as the boolean. ``1`` means the ``val``
-       should be transferred to the query sequences as its annotation;
-       ``0`` means not.
+    3. ``kingdom``. INT. The index in ``_kingdom``
 
     The table in the database file will be dropped and re-created if
     the function is re-run.
     '''
-    _overwrite(out_fp, append=append)
-    with connect(out_fp) as conn:
+    _overwrite(db_fp, **kwargs)
+    status_map = {k: i for i, k in enumerate(_status)}
+    kingdom_map = {k: i for i, k in enumerate(_kingdom)}
+    # this is the namespace for uniprot xml files.
+    ns_map = {'xmlns': 'http://uniprot.org/uniprot',
+              'xsi': 'http://www.w3.org/2001/XMLSchema-instance'}
+    n = 0
+    with connect(db_fp) as conn:
         table_name = 'metadata'
         conn.execute('''CREATE TABLE IF NOT EXISTS {t} (
                             ac       TEXT    NOT NULL,
-                            key      TEXT    NOT NULL,
-                            val      BLOB    NOT NULL,
-                            transfer BOOLEAN NOT NULL,
-                        CHECK (transfer IN (0,1)));'''.format(t=table_name))
-
-        for n, parsed in enumerate(_parse_xml(in_fp), 1):
-            insert = '''INSERT INTO {t} (ac, key, val, transfer)
-                        VALUES (?,?,?,?);'''.format(t=table_name)
-            conn.execute(insert, (parsed[0], parsed[1], parsed[2], 0))
+                            status   INT     NOT NULL,
+                            kingdom  INT     NOT NULL);'''.format(
+                                t=table_name))
+        insert = '''INSERT INTO {t} (ac, status, kingdom)
+                        VALUES (?,?,?);'''.format(t=table_name)
+
+        for fp in in_fps:
+            for i, elem in enumerate(_parse_xml(fp, ns_map), 1):
+                ac, status, kingdom = _process_entry(elem, ns_map)
```

Comment:
i c what you say. it would be impossible and nonsense if there is no
accession for an entry. it should raise exception if it should happen.
updated it.

On Thu, Feb 18, 2016 at 11:31 PM, Jamie Morton <notifications@github.com>
wrote:

> In micronota/db/uniref.py
> <https://github.com/biocore/micronota/pull/44#discussion_r53430190>:
>
> > -                            transfer BOOLEAN NOT NULL,
> > -                        CHECK (transfer IN (0,1)));'''.format(t=table_name))
> > -
> > -        for n, parsed in enumerate(_parse_xml(in_fp), 1):
> > -            insert = '''INSERT INTO {t} (ac, key, val, transfer)
> > -                        VALUES (?,?,?,?);'''.format(t=table_name)
> > -            conn.execute(insert, (parsed[0], parsed[1], parsed[2], 0))
> > +                            status   INT     NOT NULL,
> > +                            kingdom  INT     NOT NULL);'''.format(
> > +                                t=table_name))
> > +        insert = '''INSERT INTO {t} (ac, status, kingdom)
> > +                        VALUES (?,?,?);'''.format(t=table_name)
> > +
> > +        for fp in in_fps:
> > +            for i, elem in enumerate(_parse_xml(fp, ns_map), 1):
> > +                ac, status, kingdom = _process_entry(elem, ns_map)
>
> Looking here
> <https://github.com/biocore/micronota/blob/master/micronota/db/uniref.py#L367>,
> it looks like you can have accession never be set. Is this OK?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/micronota/pull/44/files#r53430190>.
>